src/java/org/apache/cassandra/index/sai/plan/Operation.java

```diff
@@ -207,7 +209,38 @@ private static int getPriority(org.apache.cassandra.cql3.Operator op)
 
     static RangeIterator buildIterator(QueryController controller)
     {
-        return Node.buildTree(controller.filterOperation()).analyzeTree(controller).rangeIterator(controller);
+        var orderings = controller.filterOperation().expressions()
+                                  .stream().filter(e -> e.operator() == Operator.ANN).collect(Collectors.toList());
+        assert orderings.size() <= 1;
+        if (controller.filterOperation().expressions().size() == 1 && orderings.size() == 1)
+        {
+            // If we only have one expression, we can use the ANN index to order and limit
+            return controller.getTopKRows(orderings.get(0));
+        }
+        else
+        {
+            var iter = Node.buildTree(controller.filterOperation()).analyzeTree(controller).rangeIterator(controller);
+            if (orderings.isEmpty())
+                return iter;
+            try
+            {
+                return controller.getTopKRows(iter, orderings.get(0));
+            }
+            finally
+            {
+                try
+                {
+                    // TODO find cleaner way to do this, but we currently need to close this later to prevent
```

Comment:
not sure wtm

Reply:
I addressed the TODO with a new iterator.